lib/swaps/Swaps.ts

```diff
@@ -185,13 +185,51 @@ class Swaps extends EventEmitter {
     this.deals.delete(deal.r_hash);
   }
 
+  /**
+   * A promise wrapper for a swap procedure
+   * @param maker the remote maker order we are filling
+   * @param taker our local taker order
+   * @returns A promise that is resolved once the swap is completed, or rejects otherwise
+   */
+  public executeSwap = (maker: StampedPeerOrder, taker: StampedOwnOrder): Promise<SwapResult> => {
+    // TODO: apply timeout as well, after implementing an interrupt signal to the swap procedure
+    return new Promise((resolve, reject) => {
+      const cleanup = () => {
+        this.removeListener('swap.paid', onPaid);
+        this.removeListener('swap.failed', onFailed);
+      };
+
+      const onPaid = (swapResult: SwapResult) => {
+        if (swapResult.r_hash === r_hash) {
+          cleanup();
+          resolve(swapResult);
+        }
+      };
+
+      const onFailed = (deal: SwapDeal) => {
+        if (deal.r_hash === r_hash) {
+          cleanup();
+          reject();
+        }
+      };
+
+      const r_hash = this.beginSwap(maker, taker);
+      if (!r_hash) {
+        reject();
```

Comment:
I think we'd want to `return` after rejecting here, so that we don't add the listeners at the end of the method.